src/main/java/uk/gov/justice/digital/zone/StructuredZoneCDC.java

```diff
@@ -0,0 +1,116 @@
+import org.apache.spark.sql.Dataset;
+import org.apache.spark.sql.Row;
+import org.apache.spark.sql.SparkSession;
+import org.jetbrains.annotations.NotNull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import uk.gov.justice.digital.config.JobArguments;
+import uk.gov.justice.digital.converter.dms.DMS_3_4_6;
+import uk.gov.justice.digital.domain.model.SourceReference;
+import uk.gov.justice.digital.exception.DataStorageException;
+import uk.gov.justice.digital.service.DataStorageService;
+import uk.gov.justice.digital.service.SourceReferenceService;
+
+import javax.inject.Inject;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.function.Consumer;
+
+import static org.apache.spark.sql.functions.col;
+import static uk.gov.justice.digital.converter.dms.DMS_3_4_6.Operation.*;
+import static uk.gov.justice.digital.converter.dms.DMS_3_4_6.ParsedDataFields.OPERATION;
+import static uk.gov.justice.digital.zone.RawZone.PRIMARY_KEY_NAME;
+
+public class StructuredZoneCDC extends StructuredZone implements DeltaWriter {
+
+    private static final Logger logger = LoggerFactory.getLogger(StructuredZone.class);
+
+    @Inject
+    public StructuredZoneCDC(
+            JobArguments jobArguments,
+            DataStorageService storage,
+            SourceReferenceService sourceReference
+    ) {
+        super(jobArguments, storage, sourceReference);
+    }
+
+    @Override
+    public Dataset<Row> process(SparkSession spark, Dataset<Row> dataFrame, Row table) throws DataStorageException {
+        val filteredRecords = dataFrame.filter(col(OPERATION).isin(cdcOperations));
+        return super.process(spark, filteredRecords, table);
+    }
+
+    @Override
+    public void writeValidRecords(
+            SparkSession spark,
+            DataStorageService storage,
+            String tablePath,
+            SourceReference.PrimaryKey primaryKey,
+            Dataset<Row> validRecords
+    ) {
+        logger.info("Applying {} CDC records to deltalake table: {}", validRecords.count(), tablePath);
+        validRecords.collectAsList().forEach(processRow(spark, storage, tablePath, primaryKey));
+
+        logger.info("CDC records successfully applied");
+        storage.updateDeltaManifestForTable(spark, tablePath);
+    }
+
+    @NotNull
+    private Consumer<Row> processRow(
+            SparkSession spark,
+            DataStorageService storage,
+            String tablePath,
+            SourceReference.PrimaryKey primaryKey
+    ) {
+        return row -> {
+            val optionalOperation = getOperation(row.getAs(OPERATION));
+            if (optionalOperation.isPresent()) {
+                val operation = optionalOperation.get();
+                try {
+                    writeRow(spark, storage, tablePath, primaryKey, operation, row);
+                } catch (DataStorageException ex) {
+                    logger.warn("Failed to {}: \n{}\nto {}", operation.getName(), row.json(), tablePath);
```

Comment:
Should this just be a warning or would error be better?

Either way, does it may sense to continue or should we abort on this exception?

Reply:
We'll revisit the log levels for the monitoring and alerting.
When a writing a CDC record fails, the streaming job should continue.